src/constraint/SequentialSampleable.cpp

```diff
@@ -0,0 +1,117 @@
+#include <aikido/constraint/SequentialSampleable.hpp>
+
+#include <dart/common/StlHelpers.hpp>
+
+namespace aikido {
+namespace constraint {
+
+// TODO (avk): Should I declare the class first and then define?
+// What is "good" C++ convention? Don't tell me it's a preference JS! -.-
+//==============================================================================
+class SequentialSampleGenerator : public SampleGenerator
+{
+public:
+  SequentialSampleGenerator(
+      statespace::StateSpacePtr stateSpace,
+      std::vector<std::unique_ptr<SampleGenerator>> generators)
+    : mStateSpace(std::move(stateSpace))
+    , mGenerators(std::move(generators))
+    , mIndex(0)
+  {
+    // Do nothing
+    // TODO (avk): Should we again check equivalence of statespaces between
+    // that of each generator and mStateSpace? I don't think it's necessary.
+  }
+
+  statespace::StateSpacePtr getStateSpace() const override
+  {
+    return mStateSpace;
+  }
+
+  bool sample(statespace::StateSpace::State* state)
+  {
+    if (!mGenerators[mIndex]->canSample())
+      mIndex++;
+
+    if (canSample())
+      return mGenerators[mIndex]->sample(state);
+
+    return false;
+  }
+
+  int getNumSamples() const override
+  {
+    return mGenerators[mIndex]->getNumSamples();
+  }
+
+  bool canSample() const override
+  {
+    if (mIndex >= mGenerators.size())
+      return false;
+
+    return true;
```

Comment:
I think this function should return true when:

1. the number of remaining generators is not zero _and_
2. one of the remaining generators is able to sample.

It seems the current implementation is missing 2.

Reply:
It sounds like `canSample()` works correctly only when it's used in `sample()`, which I don't think that's a good idea. Also, this doesn't agree with the [docstrion of `canSample()`](https://github.com/personalrobotics/aikido/blob/master/include/aikido/constraint/Sampleable.hpp#L60):
> /// Returns whether getNumSamples() > 0.

I propose changing this function to return `getNumSamples() > 0`; and changing `canSample()` not to call this function and to check the size of `mGenerators` directly in `canSample()`.

Reply:
Would it be possible for one of the remaining generators to not be sampleable? Then even though you have generators remaining (this implementation would return `true`), it still seems like you wouldn't be sample (logically, maybe it should return `false`). What if we just do something like

```
for (std::size_t i = mIndex; i < mGenerators.size(); ++i)
  if (mGenerators[i].canSample())
    return true;
return false;

Reply:
@brianhou 's version would be more efficient because returns `getNumSamples() > 0` can be more expensive as it computes the exact number of samples while @brianhou 's version can early terminate.